File: scripts/evaluator.py
import abc
import rospy
from cv_bridge import CvBridge
from sensor_msgs.msg import PointCloud2, Image
from nav_msgs.msg import Odometry
import numpy as np
import math
import matplotlib.pyplot as plt
from scripts.methods import BaseMethod
from typing import Optional, Tuple, List
import time
import os
from pathlib import Path


def angular_difference(quat_1, quat_2):
    """
    Calculate the angular difference between two quaternions.

    sql

    :param quat_1: Quaternion, the first quaternion
    :param quat_2: Quaternion, the second quaternion
    :return: float, the angle between the two quaternions in radians
    """
    cos_angle = quat_1.x * quat_2.x + quat_1.y * \
        quat_2.y + quat_1.z * quat_2.z + quat_1.w * quat_2.w
    angle = 2 * math.acos(abs(cos_angle))
    return angle


def euclidean_difference(point_1, point_2):
    """
    Calculate the Euclidean distance between two 3D points.

    sql

    :param point_1: Point, the first 3D point
    :param point_2: Point, the second 3D point
    :return: float, the Euclidean distance between the two points
    """
    dx = point_1.x - point_2.x
    dy = point_1.y - point_2.y
    dz = point_1.z - point_2.z
    dist = math.sqrt(dx*dx + dy*dy + dz*dz)
    return dist


class Evaluator(abc.ABC):
    """
    Evaluator class for evaluating loop closure detection methods.

    This class collects data, calculates real loop closures and predicted loop closures
    based on the provided methods, and calculates evaluation metrics (precision, recall, and F1 score).
    """

    def __init__(self, methods: List[BaseMethod], record_size: int, angle_consideration: bool, max_angle_deg: int, max_dist: float, min_idx_diff: int, save_candidates: bool):
        """
        Initialize the Evaluator object.

        :param methods: List[BaseMethod], list of loop closure detection methods to evaluate
        :param record_size: int, number of records to collect before evaluating
        :param angle_consideration: bool, flag to consider angular difference between poses for loop closure
        :param max_angle_deg: int, maximum angular difference in degrees for considering loop closure
        :param max_dist: float, maximum Euclidean distance for considering loop closure
        :param min_idx_diff: int, minimum index difference between two poses to consider them as potential loop closures
        """
        self.methods = methods
        self.RECORD_SIZE = record_size
        self.ANGLE_CONSIDERATION = angle_consideration
        self.MAX_ANGLE_DEG = max_angle_deg
        self.MIN_IDX_DIFF = min_idx_diff

        self.MAX_DIST = max_dist
        self.MAX_ANGLE_RAD = (self.MAX_ANGLE_DEG/180)*3.14

        self.SAVE_CANDIDATES = save_candidates

        self.odom_vec = []
        self.cloud_vec = []
        self.img_vec = []

        self.real_loop_candidates = np.full(
            (record_size, record_size), False)

        self.counter = 0
        # ros->numpy converter
        self.bridge = CvBridge()

    def synced_callback(self, cloud_msg: PointCloud2, odom_msg: Odometry, img_msg: Optional[Image]) -> None:
        """
        ROS callback function to receive synchronized sensor data and evaluate loop closure detection methods.

        :param cloud_msg: PointCloud2, point cloud message
        :param odom_msg: Odometry, odometry message
        :param img_msg: Optional[Image], image message (optional)
        :return: None
        """
        log_frequency = 50
        if self.counter % log_frequency == 0:
            rospy.loginfo("Amount of data recorded: %d", self.counter)

        # Store data to lists
        self.cloud_vec.append(cloud_msg)
        self.odom_vec.append(odom_msg.pose.pose)
        if img_msg is not None:
            image = self.bridge.imgmsg_to_cv2(
                img_msg, desired_encoding='mono8')
            self.img_vec.append(image)

        self.counter += 1

        if self.counter == self.RECORD_SIZE:

            self.get_real_loop_candidates()
            self.get_model_loop_candidates()

            # Plot odometry path
            self.plot_path()

            rospy.signal_shutdown("Evaluation completed.")

    def save_candidates(self, candidates, filename):
        if self.SAVE_CANDIDATES:
            path = Path(os.path.dirname(__file__)).parent / "results"
            if self.ANGLE_CONSIDERATION:
                path = os.path.join(path, "with_angle")
            else:
                path = os.path.join(path, "no_angle")
            
            # Check if the directory exists and create it if it doesn't
            if not os.path.exists(path):
                os.makedirs(path)
            
            path = os.path.join(path, filename)
            
            with open(path, 'w') as f:
                for i in range(1, self.RECORD_SIZE):
                    for j in range(i - self.MIN_IDX_DIFF + 1):
                        f.write(f"{int(candidates[i][j])} ")
                    f.write("\n")

    def get_real_loop_candidates(self) -> None:
        """
        Calculate real loop candidates based on Euclidean distance and angular difference between odometry poses.
        Save real loop candidates to a file if SAVE_CANDIDATES is True.

        :return: None
        """
        for i in range(1, self.RECORD_SIZE):
            for j in range(i - self.MIN_IDX_DIFF + 1):
                odom_diff = euclidean_difference(
                    self.odom_vec[i].position, self.odom_vec[j].position)
                condition = odom_diff < self.MAX_DIST

                if self.ANGLE_CONSIDERATION:
                    ang_diff = angular_difference(
                        self.odom_vec[i].orientation, self.odom_vec[j].orientation)
                    condition = condition and ang_diff < self.MAX_ANGLE_RAD

                if condition:
                    self.real_loop_candidates[i][j] = True

        method_name = str(self.methods[0]).split("_")[0]
        self.save_candidates(self.real_loop_candidates, f"real_{method_name}.txt")

    def get_model_loop_candidates(self) -> None:
        """
        Calculate predicted loop candidates using the provided loop closure detection methods.

        :return: None
        """

        best_f1_score = 0
        best_method = None
        
        for method in self.methods:
            start = time.time()
            predicted_loop_candidates = method.predict_loop_candidates(
                self.cloud_vec, self.odom_vec, self.img_vec, self.MIN_IDX_DIFF)
            duration = time.time() - start
            rospy.loginfo(f"Results for method: {method}")
            rospy.loginfo(
                f"It took {duration :.5f} seconds to predict loop candidates ({len(self.cloud_vec)} frames in total).")
            rospy.loginfo(
                f"It took {duration / len(self.cloud_vec) :.5f} seconds per frame.")


            precision, recall, f1_score, accuracy = self.calculate_metrics(predicted_loop_candidates)

            rospy.loginfo("----------------------------------------------")
            rospy.loginfo(f"Precision: {precision:.3f}")
            rospy.loginfo(f"Recall   : {recall:.3f}")
            rospy.loginfo(f"F1 Score : {f1_score:.3f}")
            rospy.loginfo(f"Accuracy : {accuracy:.3f}")
            rospy.loginfo("----------------------------------------------")
            rospy.loginfo("----------------------------------------------")
            rospy.loginfo(" ")

            if f1_score > best_f1_score:
                best_f1_score = f1_score
                best_method = method

            self.save_candidates(predicted_loop_candidates, f"{method}.txt")
        
        rospy.loginfo(f"Best method with f1_score {best_f1_score:.3f}: {best_method}")

    def calculate_metrics(self, predicted_loop_candidates: List[List[bool]]) -> Tuple[float, float, float, float]:
        """
        Calculate evaluation metrics (precision, recall, F1 score, and accuracy) based on the real and predicted loop candidates.

        :return: Tuple[float, float, float, float], evaluation metrics (precision, recall, F1 score, and accuracy)
        """
        TP = 0
        FP = 0
        FN = 0
        TN = 0

        for i in range(1, self.RECORD_SIZE):
            for j in range(i - self.MIN_IDX_DIFF + 1):
                if self.real_loop_candidates[i][j] and predicted_loop_candidates[i][j]:
                    TP += 1
                elif not self.real_loop_candidates[i][j] and predicted_loop_candidates[i][j]:
                    FP += 1
                elif self.real_loop_candidates[i][j] and not predicted_loop_candidates[i][j]:
                    FN += 1
                else:
                    TN += 1

        if TP + FP == 0:
            precision = 0
        else:
            precision = TP / (TP + FP)

        if TP + FN == 0:
            recall = 0
        else:
            recall = TP / (TP + FN)

        if precision + recall == 0:
            f1_score = 0
        else:
            f1_score = 2 * (precision * recall) / (precision + recall)

        if TP + FP + TN + FN == 0:
            accuracy = 0
        else:
            accuracy = (TP + TN) / (TP + FP + TN + FN)

        return precision, recall, f1_score, accuracy

    def plot_path(self):

        # Extract x and y coordinates from odom_vec
        x_coords = [pose.position.x for pose in self.odom_vec]
        y_coords = [pose.position.y for pose in self.odom_vec]

        # Plot the data
        plt.plot(x_coords, y_coords)
        plt.xlabel('X')
        plt.ylabel('Y')
        plt.title('Odometry Plot')
        plt.savefig("../trajectory.png")

File: src/evaluator.cpp
#include "evaluator.h"

#include <ros/ros.h>
#include "ros/package.h"

#include <fstream>
#include <cmath>
#include <chrono>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <boost/filesystem.hpp>

/**
 * \brief Calculate the angular difference between two quaternions
 *
 * \param quat_1 The first quaternion
 * \param quat_2 The second quaternion
 *
 * \return The angular difference between quat_1 and quat_2
 */
static double angular_difference(const geometry_msgs::Quaternion &quat_1, const geometry_msgs::Quaternion &quat_2)
{
    const double cos_angle = quat_1.x * quat_2.x + quat_1.y * quat_2.y + quat_1.z * quat_2.z + quat_1.w * quat_2.w;
    const double angle = 2 * acos(abs(cos_angle)); // use absolute value to avoid NaNs due to rounding errors
    return angle;
}

/**
 * \brief Calculate the Euclidean distance between two points
 *
 * \param point_1 The first point
 * \param point_2 The second point
 *
 * \return The Euclidean distance between point_1 and point_2
 */
static double euclidean_difference(const geometry_msgs::Point &point_1, const geometry_msgs::Point &point_2)
{
    const double dx = point_1.x - point_2.x;
    const double dy = point_1.y - point_2.y;
    const double dz = point_1.z - point_2.z;
    const double dist = sqrt(dx * dx + dy * dy + dz * dz);
    return dist;
}

Evaluator::Evaluator(const std::vector<BaseMethod *> &methods, int record_size, bool angle_consideration, int max_angle_deg, double max_dist, int min_idx_diff, bool save_candidates)
    : methods(methods), RECORD_SIZE(record_size), ANGLE_CONSIDERATION(angle_consideration),
      MAX_ANGLE_DEG(max_angle_deg), MAX_DIST(max_dist), MIN_IDX_DIFF(min_idx_diff), MAX_ANGLE_RAD(((double)max_angle_deg / 180) * 3.14), SAVE_CANDIDATES(save_candidates)
{
    real_loop_candidates.resize(RECORD_SIZE, std::vector<bool>(RECORD_SIZE, false));
}

void Evaluator::synced_callback(const sensor_msgs::PointCloud2::ConstPtr &cloud_msg,
                                const nav_msgs::Odometry::ConstPtr &odom_msg,
                                const boost::optional<sensor_msgs::Image::ConstPtr> &img_msg)
{
    static int counter = 0;
    static int log_frequency = 50;

    if (counter % log_frequency == 0)
    {
        ROS_INFO("Amount of data recorded: %d", counter);
    }

    // Store data to lists
    cloud_vec.push_back(*cloud_msg);
    odom_vec.push_back(odom_msg->pose.pose);
    if (img_msg)
    {
        img_vec.push_back(**img_msg);
    }

    counter++;

    if (counter == RECORD_SIZE)
    {
        get_real_loop_candidates();
        get_model_loop_candidates();

        plot_odometry_path(odom_vec);
        ros::shutdown();
    }
}

static std::string getFirstWord(const std::string &input)
{
    // Find the first space character
    size_t firstSpace = input.find_first_of(' ');

    // If there is no space character, the entire string is considered one word
    if (firstSpace == std::string::npos)
    {
        return input;
    }

    // Get the substring from the start of the string to the first space character
    return input.substr(0, firstSpace);
}

void Evaluator::save_candidates(std::vector<std::vector<bool>> &candidates, const std::string &filename){
    if (SAVE_CANDIDATES)
    {
        std::ofstream file;
        std::string path = ros::package::getPath("place_recog_eval") + "/results/";

        if (ANGLE_CONSIDERATION) {
            path += "with_angle/";
        } else {
            path += "no_angle/";
        }

        // Check if the directory exists and create it if it doesn't
        if (!boost::filesystem::exists(path)) {
            boost::filesystem::create_directories(path);
        }
        path += filename;

        file.open(path);

        for (int i = 1; i < RECORD_SIZE; i++)
        {
            for (int j = 0; j <= i - MIN_IDX_DIFF; j++)
            {
                file << candidates[i][j] << " ";
            }
            file << "\n";
        }
    }
}


void Evaluator::get_real_loop_candidates()
{
    // Calculate real loop candidates
    for (int i = 1; i < RECORD_SIZE; i++)
    {
        for (int j = 0; j <= i - MIN_IDX_DIFF; j++)
        {
            double odom_diff = euclidean_difference(odom_vec[i].position, odom_vec[j].position);
            bool condition = odom_diff < MAX_DIST;

            if (ANGLE_CONSIDERATION)
            {
                double ang_diff = angular_difference(odom_vec[i].orientation, odom_vec[j].orientation);
                condition = condition && (ang_diff < MAX_ANGLE_RAD);
            }

            if (condition)
                real_loop_candidates[i][j] = true;
        }
    }

    // Save real loop candidates to file
    std::string method_name = getFirstWord(methods[0]->getName());
    save_candidates(real_loop_candidates, "real_" + method_name + ".txt");
}

void Evaluator::get_model_loop_candidates()
{   
    double best_f1_score = 0;
    BaseMethod *best_method = nullptr;

    for (BaseMethod *method : methods)
    {
        // Predict loop candidates
        std::vector<std::vector<bool>> predicted_loop_candidates(RECORD_SIZE, std::vector<bool>(RECORD_SIZE, false));

        auto start = std::chrono::high_resolution_clock::now();
        method->predict_loop_candidates(predicted_loop_candidates, cloud_vec, odom_vec, img_vec, MIN_IDX_DIFF);
        auto duration = std::chrono::high_resolution_clock::now() - start;

        ROS_INFO("Results for method: %s", method->getName().c_str());
        ROS_INFO("It took %f seconds to predict loop candidates (%lu frames total).",
                 std::chrono::duration<double>(duration).count(), cloud_vec.size());

        ROS_INFO("It took %f seconds to predict one loop candidate.",
                 std::chrono::duration<double>(duration).count() / cloud_vec.size());

        // Calculate metrics
        double precision, recall, f1_score, accuracy;
        calculate_metrics(predicted_loop_candidates, precision, recall, f1_score, accuracy);

        // Display results
        ROS_INFO("----------------------------------------------");
        ROS_INFO("Precision: %.3f", precision);
        ROS_INFO("Recall   : %.3f", recall);
        ROS_INFO("F1 Score : %.3f", f1_score);
        ROS_INFO("Accuracy : %.3f", accuracy);
        ROS_INFO("----------------------------------------------");
        ROS_INFO("----------------------------------------------");
        ROS_INFO(" ");

        if (f1_score > best_f1_score)
        {
            best_f1_score = f1_score;
            best_method = method;
        }

        // Save predicted loop candidates to file
        save_candidates(predicted_loop_candidates, method->getName() + ".txt");
    }

    ROS_INFO("Best method: %s with f1 score: %.3f", best_method->getName().c_str(), best_f1_score);
}

void Evaluator::calculate_metrics(std::vector<std::vector<bool>> &predicted_loop_candidates, double &precision, double &recall, double &f1_score, double &accuracy)
{
    int tp = 0, fp = 0, fn = 0, tn = 0;

    for (int i = 1; i < RECORD_SIZE; i++)
    {
        for (int j = 0; j <= i - MIN_IDX_DIFF; j++)
        {
            bool predicted = predicted_loop_candidates[i][j];
            bool real = real_loop_candidates[i][j];

            if (predicted && real)
                tp++;
            else if (predicted && !real)
                fp++;
            else if (!predicted && real)
                fn++;
            else if (!predicted && !real)
                tn++;
        }
    }
    precision = (tp + fp) > 0 ? static_cast<double>(tp) / (tp + fp) : 0;
    recall = (tp + fn) > 0 ? static_cast<double>(tp) / (tp + fn) : 0;
    f1_score = (precision + recall) > 0 ? 2 * precision * recall / (precision + recall) : 0;
    accuracy = (tp + tn + fp + fn) > 0 ? static_cast<double>(tp + tn) / (tp + tn + fp + fn) : 0;
}

void Evaluator::plot_odometry_path(const std::vector<geometry_msgs::Pose> &odom_vec)
{
    // Create a blank image
    int img_size = 512;
    cv::Mat img(img_size, img_size, CV_8UC3, cv::Scalar(255, 255, 255));

    // Find min and max coordinates of the path
    double min_x = std::numeric_limits<double>::max();
    double min_y = std::numeric_limits<double>::max();
    double max_x = std::numeric_limits<double>::min();
    double max_y = std::numeric_limits<double>::min();

    for (const auto &pose : odom_vec)
    {
        min_x = std::min(min_x, pose.position.x);
        min_y = std::min(min_y, pose.position.y);
        max_x = std::max(max_x, pose.position.x);
        max_y = std::max(max_y, pose.position.y);
    }

    // Calculate scale and offsets
    double scale = std::min(img_size / (max_x - min_x), img_size / (max_y - min_y));
    double x_offset = (img_size - (max_x - min_x) * scale) / 2.0 - min_x * scale;
    double y_offset = (img_size - (max_y - min_y) * scale) / 2.0 - min_y * scale;

    // Plot each point
    cv::Point2i prev_point;
    bool prev_point_valid = false;
    for (const auto &pose : odom_vec)
    {
        cv::Point2i point(
            static_cast<int>(pose.position.x * scale + x_offset),
            static_cast<int>(img_size - (pose.position.y * scale + y_offset)));

        cv::circle(img, point, 2, cv::Scalar(0, 0, 255), -1);

        // Draw a line between the previous point and the current point, if the previous point is valid
        if (prev_point_valid)
        {
            cv::line(img, prev_point, point, cv::Scalar(255, 0, 0), 1);
        }

        // Update the previous point and set it as valid
        prev_point = point;
        prev_point_valid = true;
    }

    // Save the image to a file
    std::string path = ros::package::getPath("place_recog_eval");
    std::string filename = "odometry_path.png";
    cv::imwrite(path + "/" + filename, img);
}